### ISSUE TITLE
High CPU/memory usage in extension `<2.12` & LS `>= 0.19`

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/vscode-terraform#682]
_This is a pre-emptive explanation related to potential issues that may be caused by a planned release affecting estimated minority of users._

--- 

## Who is affected

You may experience higher than usual CPU & memory usage under the following conditions:

 - more than 1 folder (with `*.tf` files) in your workspace
 - VS Code extension version older than `2.12`
 - Language Server version equal or newer than `0.19` (released 8th July)

This is because older version of the extension never checked whether the server supports multiple folders and _always_ launched new process for each folder.

Launching 1 process per folder is a desirable behaviour when the server does not support multiple folders natively, but it if it does it means that each process would manage every single folder, thus increasing the resource usage.

`2.12.0` was released on 8th June and so this is likely to affect only a minority of users who still haven't upgraded the extension, but upgraded the language server.

## Is/was this avoidable?

We are working on reducing memory and CPU use and supporting multiple folders natively is part of it. We believe that most users will benefit from native support of multiple folders.

This extension and language server are at least two places where changes are to be made and to avoid some negative effects described above, we can wait for users to adopt the patched extension version before adopting the patched LS.
While we cannot wait forever nor expect for 100% of all users to complete the upgrade, we are making a trade-off where we can still improve the experience for the majority (which upgraded) and still ship the improvement in a timely manner.

TL;DR we left a 1 month gap between extension 2.12 (8th June 2021) and LS 0.19 (8th July 2021) as a trade-off.

## Related issues

 - https://github.com/hashicorp/vscode-terraform/issues/627
 - https://github.com/hashicorp/terraform-ls/pull/502

## How do I address this?

**Please upgrade your extension if you run into this problem**

![Screenshot 2021-07-02 at 14 35 14](https://user-images.githubusercontent.com/287584/124282734-11089900-db43-11eb-8cee-d512353480a7.png)
